File: Lab4/Lab4/extmem.cpp
/*
* extmem.c
* Zhaonian Zou
* Harbin Institute of Technology
* Jun 22, 2011
*/

#include <stdlib.h>
#include <stdio.h>
#include <cstring>
#include "extmem.h"

//初始化缓冲区，其输入参数bufSize为缓冲区大小（单位：字节），
//blkSize为块的大小（单位：字节），buf为指向待初始化的缓冲区的指针。
//若缓冲区初始化成功，则该函数返回指向该缓冲区的地址；否则，返回NULL。
Buffer *initBuffer(size_t bufSize, size_t blkSize, Buffer *buf)
{
	int i;

	buf->numIO = 0;
	buf->bufSize = bufSize;
	buf->blkSize = blkSize;
	buf->numAllBlk = bufSize / (blkSize + 1);
	buf->numFreeBlk = buf->numAllBlk;
	buf->data = (unsigned char*)malloc(bufSize * sizeof(unsigned char));

	if (!buf->data)
	{
		perror("Buffer Initialization Failed!\n");
		return NULL;
	}

	memset(buf->data, 0, bufSize * sizeof(unsigned char));
	return buf;
}

//释放缓冲区buf占用的内存空间。
void freeBuffer(Buffer *buf)
{
	free(buf->data);
}

//在缓冲区buf中申请一个新的块。若申请成功，则返回该块的起始地址；否则，返回NULL。
unsigned char *getNewBlockInBuffer(Buffer *buf)
{
	unsigned char *blkPtr;

	if (buf->numFreeBlk == 0)
	{
		perror("Buffer is full!\n");
		return NULL;
	}

	blkPtr = buf->data;

	//buf->blkSize + 1是因为每个块都有一个标志位，BLOCK_AVAILABLE块可用，BLOCK_UNAVAILABLE不可用
	while (blkPtr < buf->data + (buf->blkSize + 1) * buf->numAllBlk)
	{
		if (*blkPtr == BLOCK_AVAILABLE)//BLOCK_AVAILABLE = 0
			break;
		else
			blkPtr += buf->blkSize + 1;
	}

	*blkPtr = BLOCK_UNAVAILABLE;//BLOCK_UNAVAILABLE = 1，表示这块block已经被使用了
	buf->numFreeBlk--;
	return blkPtr + 1;//返回的是缓冲区的起始位，跳过标志位
}

//解除块blk对缓冲区内存的占用，即将blk占据的内存区域标记为可用。
void freeBlockInBuffer(unsigned char *blk, Buffer *buf)
{
	*(blk - 1) = BLOCK_AVAILABLE;
	buf->numFreeBlk++;
}

//从磁盘上删除地址为addr的磁盘块内的数据。若删除成功，则返回0；否则，返回-1。
int dropBlockOnDisk(unsigned int addr)
{
	char filename[40];

	sprintf(filename, "%d.blk", addr);
	//remove()函数用于删除指定的文件，其原型如下：
    //int remove(char * filename);
	//成功则返回0，失败则返回-1，错误原因存于errno。
	if (remove(filename) == -1)
	{
		perror("Dropping Block Fails!\n");
		return -1;
	}

	return 0;
}

//将磁盘上地址为addr的磁盘块读入缓冲区buf。
//若读取成功，则返回缓冲区内该块的地址，缓冲区buf的I/O次数加1。；
//否则，返回NULL。
unsigned char *readBlockFromDisk(unsigned int addr, Buffer *buf)
{
	char filename[40];
	unsigned char *blkPtr, *bytePtr;
	char ch;

	if (buf->numFreeBlk == 0)
	{
		perror("Buffer Overflows!\n");
		return NULL;
	}

	blkPtr = buf->data;

	while (blkPtr < buf->data + (buf->blkSize + 1) * buf->numAllBlk)
	{
		if (*blkPtr == BLOCK_AVAILABLE)
			break;
		else
			blkPtr += buf->blkSize + 1;
	}

	sprintf(filename, "%d.blk", addr);
	FILE *fp = fopen(filename, "r");

	if (!fp)
	{
		perror("Reading Block Failed!\n");
		return NULL;
	}

	//读成功后再确认申请这个块
	*blkPtr = BLOCK_UNAVAILABLE;
	blkPtr++;//移动到缓存的数据区
	bytePtr = blkPtr;

	while (bytePtr < blkPtr + buf->blkSize)
	{
		ch = fgetc(fp);
		*bytePtr = ch;
		bytePtr++;
	}

	fclose(fp);
	buf->numFreeBlk--;
	buf->numIO++;
	return blkPtr;//返回缓存的数据区
}

//将缓冲区buf内的块blk写入磁盘上地址为addr的磁盘块。
//若写入成功，则返回0；
//否则，返回-1。同时，缓冲区buf的I/O次数加1。
int writeBlockToDisk(unsigned char *blkPtr, unsigned int addr, Buffer *buf)
{
	char filename[40];
	unsigned char *bytePtr;

	sprintf(filename, "%d.blk", addr);
	FILE *fp = fopen(filename, "w");

	if (!fp)
	{
		perror("Writing Block Failed!\n");
		return -1;
	}

	for (bytePtr = blkPtr; bytePtr < blkPtr + buf->blkSize; bytePtr++)
		fputc((int)(*bytePtr), fp);

	fclose(fp);
	*(blkPtr - 1) = BLOCK_AVAILABLE;
	buf->numFreeBlk++;
	buf->numIO++;
	return 0;
}

File: DatabaseLab4/extmem.py
# __author__ = 'Xiang'
# coding:utf-8

import os


def drop_block_on_disk(addr):
    """
    从磁盘上删除地址为addr的磁盘块内的数据。
    :return:若删除成功，则返回True；否则，返回False。
    """
    filename = "%s.blk" % addr
    if os.path.exists(filename):
        os.remove(filename)
        print "删除成功"
        return True
    print "删除失败"
    return False


class Buffer:
    """
    Buffer的结构体
    """
    def __init__(self, buf_size, blk_size):
        """
        初始化
        """
        self.numIO = 0
        self.bufSize = buf_size
        self.blkSize = blk_size
        self.numAllBlk = buf_size / (blk_size + 1)
        self.numFreeBlk = self.numAllBlk
        self.Data = []
        for i in range(self.numAllBlk):
            self.Data.append([False])  # False表示块为空，True表示块中存放了数据

    def __del__(self):
        """
        析构函数
        :return:无
        """
        print "缓冲区已释放"

    def get_new_block_in_buffer(self):
        """
        在缓冲区buf中申请一个新的块。
        :return:若申请成功，则返回该块的起始地址（索引）；否则，返回NULL（-1）。
        """
        if self.numFreeBlk == 0:
            print "Buffer is full!"
            return -1

        for i in range(self.numAllBlk):
            if not self.Data[i][0]:
                self.Data[i][0] = True
                self.numFreeBlk -= 1
                return i  # 找到第一个非空的块

    def free_block_in_buffer(self, index):
        """
        解除块blk对缓冲区内存的占用，即将blk占据的内存区域标记为可用
        :return:无
        """
        self.Data[index] = [False]
        self.numFreeBlk += 1

    def read_bloc_from_disk(self, addr):
        """
        将磁盘上地址为addr的磁盘块读入缓冲区buf。
        :return:若读取成功，则返回缓冲区内该块的索引，缓冲区buf的I/O次数加1。否则，返回-1。
        """
        if self.numFreeBlk == 0:
            print "缓存区已满"
            return -1

        index = 0
        for i in range(self.numAllBlk):
            if not self.Data[i][0]:
                index = i  # 找到第一个非空的块

        filename = "%s.blk" % addr
        f = open(filename)
        if not f:
            print "打开文件失败"
            return -1

        # 成功读入后再修改
        self.Data[index] = [True]
        self.numFreeBlk -= 1
        self.numIO += 1

        data = []
        lines = f.readlines()
        for line in lines:
            data.extend(line)

        self.Data[index].append(data)
        f.close()
        return index

    def write_block_to_disk(self, addr, index):
        """
        将缓冲区buf内的块blk写入磁盘上地址为addr的磁盘块。
        :return:若写入成功，则返回True,同时,缓冲区buf的I/O次数加1；否则，返回False。
        """
        filename = "%s.blk" % addr
        f = open(filename, 'w')

        if not f:
            print "打开文件失败"
            return False

        f.writelines(self.Data[index][1])
        f.close()
        self.Data[index] = [False]  # 写入后该块释放
        self.numFreeBlk += 1
        self.numIO += 1
        return True


if __name__ == '__main__':
    a = ['1\n', '2', '3', '4']
    b = open('1.txt', 'w')
    b.writelines(a)

